kubespawner/utils.py

```diff
@@ -222,3 +223,45 @@ def recursive_update(target, new):
 
         else:
             target[k] = v
+
+
+class IgnoreMissing(dict):
+    """
+    Dictionary subclass for use with format_map
+
+    Renders missing values as "{key}", so format strings with
+    missing values just get rendered as is.
+
+    Stolen from https://docs.python.org/3/library/stdtypes.html#str.format_map
+    """
+
+    def __missing__(self, key):
+        return f"{{{key}}}"
+
+
+def recursive_format(
+    format_object: Union[str, dict, list], **kwargs
+) -> Union[str, dict, list]:
+    """
+    Recursively format given object with values provided as keyword arguments.
+
+    If the given object (string, list or dict) has items that do not have
+    placeholders for passed in kwargs, no formatting is performed.
+
+    recursive_format("{v}", v=5) -> Returns "5"
+    recrusive_format("{a}") -> Returns "{a}" rather than erroring, as is
+    the behavior of "format"
+    """
+    if isinstance(format_object, str):
+        return format_object.format_map(IgnoreMissing(kwargs))
+    elif isinstance(format_object, list):
+        return [recursive_format(i, **kwargs) for i in format_object]
```

Comment:
Not really important, since this is probably not getting called a million times, but it would be a bit more efficient to create the IgnoreMissing once and pass it around like `recursive_format(obj, format_dict)`, rather than create a fresh kwargs dict (calling a function with `**kwargs` creates a new dict to be passed) and IgnoreMissing for every value.

Reply:
@minrk I want the call sites to still be able to pass in values as **kwargs, rather than have to import and pass in IgnoreMissing. Think it's ok to leave this as is?

Reply:
I do, it's not a big deal. I was thinking the function would expect a dict (like `format_map`), but we'd cast to IgnoreMissing if it wasn't cast already, so the cast would happen at most once, rather than N times, but still callers wouldn't need to know anything about IgnoreMissing.